### ISSUE TITLE
Scale test (adding 40 nodes) failed by timeouts(nodes added to long) and errors on teardown: seastar::rpc::remote_verb_error (Mutation of 17696794 bytes is too large for the maximum size of 16777216)

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylladb#17545]
Scylla version 5.5.0~dev-0.20240221.4be70bfc2bc7 with build-id be290d67e3f77a2c2c30c7de606fbca196424c97

Large scale test: [update_cluster_layout_tests.py::TestLargeScaleCluster::test_add_many_nodes_under_load](https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/79/artifact/logs-full.release.001/1708686228808_update_cluster_layout_tests.py%3A%3ATestLargeScaleCluster%3A%3Atest_add_many_nodes_under_load) try to add 40 nodes. But looks like it failed by timeout because node25 bootstrapped too long and dtest found on teardown a lot of errors:
```
['ERROR 2024-02-23 10:55:47,702 [shard 0:strm] raft - [ae8ff257-e455-4ebe-913e-45bf24116628] Transferring snapshot to 7dde096b-b9f8-428f-9339-15288e5970ed failed with: seastar::rpc::remote_verb_error (Mutation of 17696794 bytes is too large for the maximum size of 16777216)', 'ERROR 2024-02-23 10:56:06,597 [shard 0:strm] raft - [ae8ff257-e455-4ebe-913e-45bf24116628] Transferring snapshot to 7dde096b-b9f8-428f-9339-15288e5970ed failed with: seastar::rpc::remote_verb_error (Mutation of 17696794 bytes is too large for the maximum size of 16777216)', 'ERROR 2024-02-23 10:53:28,396 [shard 0:strm] raft - [ae8ff257-e455-4ebe-913e-45bf24116628] Transferring snapshot to 7dde096b-b9f8-428f-9339-15288e5970ed failed with: seastar::rpc::remote_verb_error (Mutation of 17696794 bytes is too large for the maximum size of 16777216)'
```

Logs: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/79/artifact/logs-full.release.001/

**Comment by aleksbykov** [comment]
@kbr-scylla can you take a look

**Comment by gusev-p** [comment]
`ERROR 2024-03-01 15:59:14,423 [shard 0:stmt] group0_raft_sm - group0_state_machine::transfer_snapshot, merge_topology_snapshot CRASH std::invalid_argument (Mutation of 17698243 bytes is too large for the maximum size of 16777216), topo_desc {1, 408961}, cdc_gen_desc {1, 17280322}, topo_req_desc {25, 19536}`

So we are crashing because of a huge `cdc_generations_v3` mutation. 

The strange thing is that in `prepare_new_cdc_generation_data` we carefully limit individual mutation sizes by `_raft.max_command_size()` within one generation, but in `merge_topology_snapshot` we write the entire `cdc_generations_v3` table, including all the previous generations, as a single mutation. @kbr-scylla maybe I miss something here.

**Comment by kbr-scylla** [comment]
@gusev-p your assessment looks correct to me. Bug in topology snapshot transfer / merging.

BTW. why didn't we catch this in the test.py 100 node test? I'm guessing it's because we're limiting vnodes so generations are smaller? Would be good if we had a reproducer in test.py too, with small number of nodes preferably, maybe by bloating the size of generation mutations artificially or sth

**Comment by gusev-p** [comment]
> @gusev-p your assessment looks correct to me. Bug in topology snapshot transfer / merging.

How should we fix that? `segment_manager::allocate_when_possible` requires that the total size of all mutations, passed into `_db.local().apply` be less than `max_mutation_size`, they are written atomically. Could we split this to multiple `apply`-s? There is a TODO
`// TODO ensure atomicity of snapshot application in presence of crashes (see TODO in `apply`)`
but I don't what it refers to.

> Would be good if we had a reproducer in test.py too, with small number of nodes preferably, maybe by bloating the size of generation mutations artificially or sth

Yes, ok

**Comment by kbr-scylla** [comment]
The CDC_GENERATIONS_V3 mutation could easily be split into multiple applys without breaking any kind of atomicity/isolation. The data is not being referred/used by anything until corresponding "pointer" is set in system.topology table (i.e. added to `committed_cdc_generations` and maybe `unpublished_cdc_generations`).

So while the system.topology table mutations should probably be applied atomically (otherwise, after restart, `topology_state_load` may observe some partial incorrect state and do who knows what), the cdc_generations_v3 table mutations can be applied partially (assuming we apply them before the system.topology mutations that depend on them)

**Comment by mykaul** [comment]
Are we going to handle this for 6.0?

**Comment by kbr-scylla** [comment]
Of course. It's a blocker.

**Comment by mykaul** [comment]
> Of course. It's a blocker.

It is now.

**Comment by kbr-scylla** [comment]
@aleksbykov please retest on latest version

**Comment by aleksbykov** [comment]
@kbr-scylla , 1 time test passed. but the latest runs always failes. some times with same errors, sometimes with errors:

and again not all nodes are added to cluster
```
'node1', ['ERROR 2024-03-26 13:09:40,849 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:10:17,462 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:10:22,267 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:09:47,848 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:10:19,461 [shard 0:strm]
```
Job: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/213/testReport/
Scylla version: 5.5.0~dev-0.20240325.9789a3dc7c0f with build-id 72988c2aa0fdc2ce972d029228f9d211ed70ee3c
i checked, the commit with fix should be in  for this build

**Comment by kbr-scylla** [comment]
Thanks. Moving to backlog again

**Comment by kbr-scylla** [comment]
So @aleksbykov are you saying that test was passing for some time, but then there was a regression?

**Comment by gusev-p** [comment]
This is another problem, will look at it

**Comment by aleksbykov** [comment]
> So @aleksbykov are you saying that test was passing for some time, but then there was a regression?

yes, i pretty sure, that i run 1 or 2 times the test and it passed. Then i prepare the PR to enable the test for regular runs, and it start failed.

**Comment by kbr-scylla** [comment]
> > So @aleksbykov are you saying that test was passing for some time, but then there was a regression?
> 
> yes, i pretty sure, that i run 1 or 2 times the test and it passed. Then i prepare the PR to enable the test for regular runs, and it start failed.

Was Raft snapshot transfer extended with some new data in the last ~week?
Tablets? Auth? Service levels?
@tgrabiec @bhalevy @nuivall @Jadw1 @piodul

**Comment by kbr-scylla** [comment]
Could also be that before, snapshot transfer was retrying indefinitely and for some reason, would eventually succeed...
But now we have a timeout on node bootstrap read barrier:
```
('node36', ['ERROR 2024-03-26 13:10:37,917 [shard 0:main] init - Startup failed: service::raft_operation_timeout_error (group [58723141-eb6f-11ee-a911-2feb07326972] raft operation [read_barrier] timed out)'])
```

**Comment by avikivity** [comment]
Isn't cdc_generations_v3 a raft table? Shouldn't it be using the schema commitlog?

And I still think we need the ability for commits to span multiple commitlog segments.

**Comment by nuivall** [comment]
> Was Raft snapshot transfer extended with some new data in the last ~week?

yes, service levels were added by Michal (supposed to be small size)

**Comment by avikivity** [comment]
> > Isn't cdc_generations_v3 a raft table? Shouldn't it be using the schema commitlog?
> 
> It is and it is

Then why the error with 16MB threshold? schema commitlogs are larger. Maybe I was looking at an older comment.

**Comment by gusev-p** [comment]
> > > Isn't cdc_generations_v3 a raft table? Shouldn't it be using the schema commitlog?
> > 
> > 
> > It is and it is
> 
> Then why the error with 16MB threshold? schema commitlogs are larger. Maybe I was looking at an older comment.

hm, I use the wrong commitlog when I split mutations in `merge_topology_snapshot` (and in `cdc_generation_mutations_topology_snapshot_replication` injection), but it shouldn't hurt since schema_commitlog_segment_size_in_mb > commitlog_segment_size_in_mb

**Comment by aleksbykov** [comment]
Next run was succesfull: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/199
then next 2 runs failed:
Run without raft topology: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/202/testReport/
Run with raft topology: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/201/

The run 201 with raft topology failed with next errors:
```
 gossip - gms/gossiper.cc(735:32) `future<> gms::gossiper::do_status_check()`: waiting for endpoint lock (ep=127.0.45.26) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(568:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(568:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`']), ('node7', ['ERROR 2024-03-25 14:34:13,008 [shard 0: gms] gossip - gms/gossiper.cc(568:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`: waiting for endpoint lock (ep=127.0.45.18) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(568:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(568:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`']), ('node9', ['ERROR 2024-03-25 14:36:21,929 [shard 0: gms] gossip - Gossip change listener failed: std::runtime_error (Host 127.0.45.28 does not have HOST_ID application_state), at: 0x63472fe 0x6347910 0x6347bf8 0x5dff8be 0x5dffa77 0x3e094bd 0x141cf7a 0x5e3fb8f 0x5e40e77 0x5e401d8 0x5dce5f7 0x5dcd7ac 0x13ada19 0x13af470 0x13ac09c /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/6bdb456fadfef65d9719069a5f43ad376ffa9d2d/libreloc/libc.so.6+0x27b89 /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/6bdb456fadfef65d9719069a5f43ad376ffa9d2d/libreloc/libc.so.6+0x27c4a 0x13a9924']), ('node10', ['ERROR 2024-03-25 14:36:08,393 [shard 0: gms] gossip - gms/gossiper.cc(568:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`: waiting for endpoint lock (ep=127.0.45.24) took more than 1m, signifying
 ```
 
 the latest run: https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/Dtest/job/master-dtest-with-raft/213/testReport/
 
 failed with errors:
 ```
 ('node1', ['ERROR 2024-03-26 13:09:40,849 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:10:17,462 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:10:22,267 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:09:47,848 [shard 0:strm] raft - [3263c758-e3b0-4982-94e4-cdc1be8a887c] Transferring snapshot to 9ac8ebc0-3dbf-42cf-adfc-0e9524e79be7 failed with: seastar::rpc::remote_verb_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)', 'ERROR 2024-03-26 13:10:19,461 [shard 0:strm]
 ```

**Comment by gleb-cloudius** [comment]
Well, it looks like now simply querying data during snapshot transfer  (probably CDC generations)  fails because unpaged query it too large. The limit is 100MB.

**Comment by kbr-scylla** [comment]
No. For example, if we created e.g. 3 generations in the span of a minute, then we need to transfer all 3.

We have a mechanism to garbage-collect generations after they were closed. But we do that only after 24h.
See 3ee56e1936a54be1927b810c6f4ca89c293c7d1f where this is explained in detail

We could potentially reduce the 24h. But it's still all timing based, if we boot nodes in quick succession, we may create multiple generations that all need to be transferred. So while in practice where node bootsraps are rare the number of generations will be small most of the time, in theory the number can be large.

We should handle this during snapshot transfer IMO. Paged queries and/or streaming the snapshot data etc. whatever's necessary to handle large snapshots.

**Comment by gleb-cloudius** [comment]
Well, this CDC generation thing is complete clusterf*ck TBH. Especially after the patch that allows writing to an old generations. More than 100MB data for 36 nodes? Crazy. We need to reduce it somehow and not get into the project of new protocol for snapshot transfer before 6.0.

**Comment by kbr-scylla** [comment]
Then let's reduce the 24h. Here's the explanation:
```cpp
        // If some node can still accept a write to a CDC generation, we cannot delete this generation.
        // A request coordinator accepts a write to one of the previous generations (the ones that stopped
        // operating before `now`) if the write's timestamp is higher than `now - 5s` where `now` is
        // provided by the coordinator's local clock. So, we can safely delete a generation if it
        // stopped operating more than 5 s ago on all nodes. Since their clocks can be desynchronized,
        // we delete only generations that stopped operating more than 24 hours ago, which is a safe and
        // sufficient choice.
        auto ts_upper_bound = db_clock::now() - std::chrono::days(1);
        utils::get_local_injector().inject("clean_obsolete_cdc_generations_change_ts_ub", [&] {
            ts_upper_bound = db_clock::now();
        });
```

since we only want to account for clock desync really, how about using something like `ring_delay + generation_leeway`? I.e. we assume that clocks are not more desynced than `ring_delay + generation_leeway`.

cc @patjed41

**Comment by patjed41** [comment]
> Then let's reduce the 24h. Here's the explanation:
> 
> ```c++
>         // If some node can still accept a write to a CDC generation, we cannot delete this generation.
>         // A request coordinator accepts a write to one of the previous generations (the ones that stopped
>         // operating before `now`) if the write's timestamp is higher than `now - 5s` where `now` is
>         // provided by the coordinator's local clock. So, we can safely delete a generation if it
>         // stopped operating more than 5 s ago on all nodes. Since their clocks can be desynchronized,
>         // we delete only generations that stopped operating more than 24 hours ago, which is a safe and
>         // sufficient choice.
>         auto ts_upper_bound = db_clock::now() - std::chrono::days(1);
>         utils::get_local_injector().inject("clean_obsolete_cdc_generations_change_ts_ub", [&] {
>             ts_upper_bound = db_clock::now();
>         });
> ```
> 
> since we only want to account for clock desync really, how about using something like `ring_delay + generation_leeway`? I.e. we assume that clocks are not more desynced than `ring_delay`.
> 
> cc @patjed41

We can do it and I think `ring_delay + generation_leeway` is a good choice. I'm not sure how much it will help though. We can still create multiple generations within `ring_delay + generation_leeway`.

**Comment by gleb-cloudius** [comment]
> > Then let's reduce the 24h. Here's the explanation:
> > ```c++
> >         // If some node can still accept a write to a CDC generation, we cannot delete this generation.
> >         // A request coordinator accepts a write to one of the previous generations (the ones that stopped
> >         // operating before `now`) if the write's timestamp is higher than `now - 5s` where `now` is
> >         // provided by the coordinator's local clock. So, we can safely delete a generation if it
> >         // stopped operating more than 5 s ago on all nodes. Since their clocks can be desynchronized,
> >         // we delete only generations that stopped operating more than 24 hours ago, which is a safe and
> >         // sufficient choice.
> >         auto ts_upper_bound = db_clock::now() - std::chrono::days(1);
> >         utils::get_local_injector().inject("clean_obsolete_cdc_generations_change_ts_ub", [&] {
> >             ts_upper_bound = db_clock::now();
> >         });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > since we only want to account for clock desync really, how about using something like `ring_delay + generation_leeway`? I.e. we assume that clocks are not more desynced than `ring_delay`.
> > cc @patjed41
> 
> We can do it and I think `ring_delay + generation_leeway` is a good choice. I'm not sure how much it will help though. We can still create multiple generations within `ring_delay + generation_leeway`.

So who are going to do it. I'd rather not touch CDC :).

**Comment by bhalevy** [comment]
> > @[annastuchlik](https://github.com/annastuchlik) annastuchlik closed this as [completed](https://github.com/scylladb/scylladb/issues?q=is%3Aissue+is%3Aclosed+archived%3Afalse+reason%3Acompleted) in [annastuchlik/scylla-enterprise@4dfb7e3](https://github.com/annastuchlik/scylla-enterprise/commit/4dfb7e30513c30ae101e69f34004a4ee5d48aad8) [Apr 23, 2024](https://github.com/scylladb/scylladb/issues/17545#event-12574964768)
> 
> lol
> 
> I don't understand what was the actual event that caused the close -- was there a merge in this fork? Or creation of fork itself?

@annastuchlik did you push the above patch to an "enterprise" branch in your repo?

**Comment by kbr-scylla** [comment]
The automation did not move it from `Done` even though I reopened the issue :( and I forgot about it for a while because of that

@aleksbykov please rerun the scale test on latest build. There were some fixes from @gleb-cloudius and @patjed41 for issues detected in your runs

**Comment by bhalevy** [comment]
> The automation did not move it from `Done` even though I reopened the issue :( and I forgot about it for a while because of that

https://github.com/scylladb/scylladb/pull/18497 just `Refs` this issue.  It doesn't "Fix" it with the right github keyword.

**Comment by kbr-scylla** [comment]
> https://github.com/scylladb/scylladb/pull/18497 just Refs this issue. It doesn't Fix it with the right github keyword.

My problem is not about moving it **to** Done, but **from** Done. It was mistakenly moved to Done before, and then when I reopened the issue, it was not moved back. Causing me to forget about it.

(I modified the GH project workflow now to move issues back to backlog when reopening.)

**Comment by mykaul** [comment]
@aleksbykov - can we re-test this? Now that schema log is bigger, we should see better results. Not fixed yet, but improved situation (which will allow us to remove it from the release blockers list)

**Comment by aleksbykov** [comment]
@kbr-scylla @mykaul , Sorry for delay, we have long holidays. 
Anyway, i ran several times the test and it passed with consistent-topology-changes feature

**Comment by mykaul** [comment]
> @kbr-scylla @mykaul , Sorry for delay, we have long holidays. Anyway, i ran several times the test and it passed with consistent-topology-changes feature

Passed as in how many nodes?

**Comment by aleksbykov** [comment]
40 nodes

**Comment by mykaul** [comment]
> 40 nodes

Thanks. Next we should test 100.

**Comment by mykaul** [comment]
https://github.com/scylladb/scylladb/pull/17632 was marked as fixes this issue, so I'm not sure why it's still open. I'm sure we have tasks here, but perhaps they should be in a separate issue.

**Comment by bhalevy** [comment]
> #17632 was marked as fixes this issue, so I'm not sure why it's still open. I'm sure we have tasks here, but perhaps they should be in a separate issue.

Yes.  This issue is about a specific error and if it was addressed we should close this issue and open new one(s) if other failures are found.

**Comment by bhalevy** [comment]
> #17632 was marked as fixes this issue, so I'm not sure why it's still open.

[kbr-scylla](https://github.com/kbr-scylla) reopened this [on Mar 26](https://github.com/scylladb/scylladb/issues/17545#event-12249998683)

> I'm sure we have tasks here, but perhaps they should be in a separate issue.

**Comment by kbr-scylla** [comment]
>  Anyway, i ran several times the test and it passed with consistent-topology-changes feature

Great, thanks!

> https://github.com/scylladb/scylladb/pull/17632 was marked as fixes this issue, so I'm not sure why it's still open. I'm sure we have tasks here, but perhaps they should be in a separate issue.

If the test is passing, the issue is fixed. Previously I reopened because the PR mistakenly closed it -- the test wasn't passing.

> Thanks. Next we should test 100.

Agreed. @aleksbykov can you please also run a similar scale test but with 100 nodes?

This particular issue I am finally closing.

**Comment by aleksbykov** [comment]
> > 40 nodes
> 
> Thanks. Next we should test 100.

@kbr-scylla @mykaul 
with one hundred nodes test failed on node 72: with startup failed:
```
INFO  2024-05-14 09:35:44,350 [shard 0:main] init - Shutting down sighup was successful
INFO  2024-05-14 09:35:44,350 [shard 0:main] init - Shutting down configurables
INFO  2024-05-14 09:35:44,350 [shard 0:main] init - Shutting down configurables was successful
ERROR 2024-05-14 09:35:44,350 [shard 0:main] init - Startup failed: std::runtime_error (Bootstrap failed. See earlier errors (Rolled back: Failed to commit cdc generation: std::runtime_error (raft topology: exec_global_command(barrier) failed with seastar::rpc::closed_error (connection is closed))))
```
Node1 reported:
```
INFO  2024-05-14 09:35:09,187 [shard 0:main] raft_group_registry - marking Raft server 653fc7f9-529e-4226-a3a2-c59480e630ef as alive for raft groups
WARN  2024-05-14 09:35:09,217 [shard 0:stmt] mutation_partition - Memory usage of unpaged query exceeds soft limit of 1048576 (configured via max_memory_for_unlimited_query_soft_limit)
WARN  2024-05-14 09:35:09,231 [shard 0:stmt] mutation_partition - Memory usage of unpaged query exceeds soft limit of 1048576 (configured via max_memory_for_unlimited_query_soft_limit)
INFO  2024-05-14 09:35:10,057 [shard 0:strm] boot_strapper - Get random bootstrap_tokens={1433426564769120107, -3024130438421483493, 6783300416989774946, -1913632903514385366, -1296153106641068349, 1007429126839380150, 7984639701427359074, -5786157125825623110, -7359984701809794846, 3522986124224876937, -3441361695749005185, -1206305428627841913, -3124027604652700064, 2725068477338432786, 5205155087332034779, 2433256482079879044, 3679214181567140968, -5816330808577396288, -118004484174318508, 9039937626987112835, -8986726699696615599, -2490997589493233577, -7403114531822865343, -5716614150496756632, -7493452086901961486, -2617472633652495306, -3959849680369634054, -6734145791635177944, 3700399166162257380, 4663913147842742309, -7037273237366318004, -3628651639882969965, -7687167558551732292, -4990535938613735299, 5358693883544366981, -4977344577775304537, -8589655038447844655, 7083891868698313787, -8863804598399540463, -8771150581826651932, -5767987021411659170, 5218986179965462847, 2563434673873028804, 7162972308007128904, -8252511960786747378, 6376868954473548859, -76401018025807680, 2710714891402271653, 1656538783571724615, 418032343716706861, 6697443819176173875, -3514937049858708671, -1863753423742103941, -6400004109999888752, -2692834300114109558, 1772721558144266394, -2575621367426696002, 383763399075815123, 214228935000835408, -6262633267021563247, 1690187744831784835, -3384833187351254953, 1891341548161072751, 2528595013806715080, -1284056852921209896, 5133445597581039797, 6019522473871943281, 7315350882739636578, 9051377965348576980, -1397375351519384899, -6205047198403160958, 6859895283439108025, 7954942467421122994, -7872251089852618669, 8571203620064791618, -3677958235254371874, -6859785428053165913, 2282647432688241035, 1800670181316544576, -1859478944350163913, -2250514226640233767, -7938529939655239330, 4206354305925898378, -2937263044357094893, -5008816272078468082, -5498669479214528962, -5958746016302597292, -8971552293989757935, -9106251332869317242, 5918305565532252436, -3841807879079826123, -7323785487124108681, -7735831486495017644, 7204087527042597046, -7232993587521638169, -4755186106873936807, -958779569968857377, 1916323627092394585, 7212456190278428228, 7157569241964554297, -6515071474984575560, -7911708490078297653, -8150780494158497303, -6323264798375069181, -5039969876466999202, 8169410260307209331, -8101573313025566123, 825092948703179705, 5122021485732998520, 724335777260500712, 1572630048126628947, 625943744704382258, -5729509194927664742, -945883982220334658, -3266957823789072687, 3200194487146988246, 4953675008171200170, 5956725420241501323, 4357968297482089038, -8113442556620021593, -5462829191312072664, 3523901485555516340, -2603353483898286824, 3932941005767061773, -6556989695979498498, 7757339169541007516, 8818407673633376845, 1896016089707419861, -2797834054124500170, -7156250960156119166, 632977869553601190, 4424575218781430862, 3758371095514587868, -4528231696225597070, -8077240885193817196, -676982939146095600, 32681540223590446, -7711062772471220431, -3152853350762551531, 7725397093181086598, -3161376335528435746, 7771306044333077868, -3821692663245420267, -6306188336484368604, -3094319437721657761, -802116362462267891, 9214007929849164490, 116433898980682754, 7697789486638277069, -4075465411742528477, -7326517739065625923, 2112368831788757319, -6170387378614800072, 2253725650358131408, 6700888590293025599, 2168865430477276424, 695649264342536306, -7971750163187343559, -3170733055132740834, 7861011416081269000, -2273925284435075797, 5114868705041233173, 5564589071519102085, 6713810811342183230, -356363176651482907, 4275472870740204854, 5697228932344174734, 6691864265103870369, 625188329207482295, -868587851645524464, 7996393205484900109, -390104312926003599, 1328171356989826527, -974771486148213499, 6091405376432035272, -6433226178234882108, 7593238809332649617, -8109742698956844364, 4646230940499213604, -9117438798757406119, 3991923048465018930, -8227945738875681247, -1253549626841561311, -7875609559368782488, 2859925769821374569, -4086010592457615043, 4574271297010416851, 7018320147844907384, 3047475207172594410, -3154963074535217256, -3444347913241611466, 4042803308545958286, 2664649119029700058, 5091259864602170296, 6602591859512010379, 5185395581325197002, 1150375261101609956, -3501784990426085068, -1337623305353950550, -9122397680101266306, 5163435014056426504, -7303688106280964990, -8713058669476528687, -3505071432777038461, 3858149921735920, -3825907779282338016, 76516151967863266, -4836007009399991140, -4950983275267961235, -4336198082505719916, 6119017763310869985, 6741462379366766122, 1976938571152401547, -401356364890233257, 7350557784710934185, 1502329198412860627, -1376843198472012287, 6458266078616169864, 25452585543239930, -7562222136015234239, -3745334077680945772, 6721213754361735353, -6882100446714222226, 7611898378492396288, -7585761129002046485, 8559125686671297708, 5711293982266954434, -2619363999579083185, 3484361002832311692, 8019752824201374624, 1243571285313864769, 5317769688969194095, -3994366891095412569, -9125477489958615233, 5298270704832328366, 4728581015820436936, -3441737662948346413, 7216465223581840673, -7869446220183283232, 5500695658777323719, 5211148161944101124, 6445779647686489488, 6729456479580983997, -8760014131959324523, -7673231066596070984, 2136180900933206657, 660876118807753351, 4194089243637930954, 3434951662449236451, -7144918282434882076, -1629403148235721976, -2875386342848361916, 2601130810192064008, 5832775409262545507, 4633047589837472532, 8995914355180526706}
INFO  2024-05-14 09:35:10,211 [shard 0:strm] raft_topology - updating topology state: bootstrap: insert tokens and CDC generation data (UUID: 41bfd970-11d5-11ef-59b4-87d0a0ed676e)
ERROR 2024-05-14 09:35:10,267 [shard 0: gms] lsa - Aborting due to allocation failure: failed to refill emergency reserve of 30 (have 12 free segments)
Aborting on shard 0.
Backtrace:
  0x5eb6d48
  0x5eed5a1
  /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/libreloc/libc.so.6+0x3dbaf
  /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/libreloc/libc.so.6+0x8e883
  /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/libreloc/libc.so.6+0x3dafd
  /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/libreloc/libc.so.6+0x2687e
  0x213a667
  0x1d995bd
  0x1c4cf28
  0x1b19ad6
  0x1ba5fb7
  0x143784a
  0x5ec897f
  0x5ec9c67
  0x5ec8fc8
  0x5e56ed7
  0x5e5609c
  0x13c7cb8
  0x13c9700
  0x13c62cc
  /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/libreloc/libc.so.6+0x27b89
  /jenkins/workspace/scylla-staging/abykov/Dtest/master-dtest-with-raft/scylla/.ccm/scylla-repository/2ce643d06bf04269661c1612b9b209a6ecc2a1b1/libreloc/libc.so.6+0x27c4a
  0x13c37e4
 ```

```
[Backtrace #0]
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:68
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:825
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:855
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:867
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:4071
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:4047
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:4043
/data/scylla-s3-reloc.cache/by-build-id/2e2c89cbb469c1231861753c4af823040a31579e/extracted/scylla/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=70e92bb237883be3065a6afc9f0696aef2d068bf, for GNU/Linux 3.2.0, not stripped

__GI___sigaction at :?
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
logalloc::allocating_section::reserve(logalloc::tracker::impl&) at ./utils/logalloc.cc:2945
decltype(auto) logalloc::allocating_section::with_reserve<logalloc::allocating_section::operator()<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}>(logalloc::region&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}&&)::{lambda()#1}>(logalloc::region, logalloc::allocating_section::operator()<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}>(logalloc::region&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}&&)::{lambda()#1}) at ././utils/logalloc.hh:473
 (inlined by) decltype(auto) logalloc::allocating_section::operator()<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}>(logalloc::region&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}&&) at ././utils/logalloc.hh:529
 (inlined by) operator() at ./replica/memtable.cc:800
 (inlined by) decltype(auto) with_allocator<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0>(allocation_strategy&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0&&) at ././utils/allocation_strategy.hh:318
 (inlined by) replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&) at ./replica/memtable.cc:799
void replica::table::do_apply<frozen_mutation const&, seastar::lw_shared_ptr<schema const>&>(replica::compaction_group&, db::rp_handle&&, frozen_mutation const&, seastar::lw_shared_ptr<schema const>&) at ./replica/table.cc:2816
 (inlined by) operator() at ./replica/table.cc:2844
 (inlined by) seastar::future<void> seastar::futurize<void>::invoke<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&>(replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&) at ././seastar/include/seastar/core/future.hh:2032
 (inlined by) auto seastar::futurize_invoke<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&>(replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&) at ././seastar/include/seastar/core/future.hh:2066
 (inlined by) seastar::futurize<std::result_of<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0 ()>::type>::type replica::dirty_memory_manager_logalloc::region_group::run_when_memory_available<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0>(replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ./replica/dirty_memory_manager.hh:572
 (inlined by) replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ./replica/table.cc:2843
replica::database::apply_in_memory(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ./replica/database.cc:1833
replica::database::do_apply(seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, seastar::bool_class<db::force_sync_tag>, std::variant<std::monostate, db::per_partition_rate_limit::account_only, db::per_partition_rate_limit::account_and_enforce>) at ./replica/database.cc:2053
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/coroutine:240
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:125
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2690
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:3152
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3320
seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3210
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ./build/release/seastar/./seastar/src/core/app-template.cc:276
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at ./build/release/seastar/./seastar/src/core/app-template.cc:167
scylla_main(int, char**) at ./main.cc:682
std::function<int (int, char**)>::operator()(int, char**) const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
main at ./main.cc:2172
__libc_start_call_main at ??:?
__libc_start_main_alias_2 at :?
_start at ??:?

```
The test is same as 40 nodes. But also problem could be with instance where it run, instance could don't have enough resources for 1 hundred nodes.

**Comment by kbr-scylla** [comment]
Thanks @aleksbykov.

> But also problem could be with instance where it run, instance could don't have enough resources for 1 hundred nodes.

I doubt it. It looks like the problem happens when we try to update metadata. Metadata shouldn't be too large even with 100 nodes. I wonder which exact write causes this crash. Could be many things. For example it could happen when we try to write to Raft log table.

Anyway we can verify that it's a regression --- @aleksbykov can you also run the test with 100 nodes on 5.4 and see if it also crashes in similar way?

In any case, it looks like a critical issue to me and we must investigate. @aleksbykov please open a new issue for 100 nodes so we don't keep adding to this issue, and assign to me. (After also checking if it happens for 5.4 -- I think it's an important bit of information)

**Comment by aleksbykov** [comment]
@kbr-scylla , 
WDYT, should be test run also with --force-gossip-topology-changes on master?

**Comment by kbr-scylla** [comment]
> I doubt it. It looks like the problem happens when we try to update metadata. Metadata shouldn't be too large even with 100 nodes. I wonder which exact write causes this crash. Could be many things. For example it could happen when we try to write to Raft log table.

Actually now I remembered that this is a **dtest**, so all nodes are running on a single machine. So it could be indeed the instance size as you said. But we should check if gossip mode doesn't have this problem too.

> @kbr-scylla ,
> WDYT, should be test run also with --force-gossip-topology-changes on master?

Good idea. We could compare 3 runs:
- master with `--force-gossip-topology-changes`
- master with raft-topology (we already have this and know it breaks)
- 5.4

**Comment by kbr-scylla** [comment]
Also, how many shards is this using? Is it `--smp 2` (IIRC the default with dtest)? Is it running with `--overprovisioned`?

**Comment by aleksbykov** [comment]
I opened issue: https://github.com/scylladb/scylladb/issues/18669. Continue to update status there

**Comment by aleksbykov** [comment]
> Also, how many shards is this using? Is it `--smp 2` (IIRC the default with dtest)? Is it running with `--overprovisioned`?

yes: 
` smp, (positional) 2, memory, (positional) 1024M,`
